### PR TITLE
DOP-3969: Raise a diagnostic for section structures that cannot fit into a tree

### PR DESCRIPTION
**Pull request by i80and**
Our approach here is to add a stack of "minimum heading levels" to our parsing memo. When we enter and exit a directive, we push and pop the current heading level, respectively.

This allows us to warn when we encounter a section with an unacceptably low section level.

### CODE REVIEW

File: snooty/tinydocutils/states.py [hunk 1]
```diff
@@ -1103,6 +1109,14 @@ def check_subsection(self, source: str, style: StyleKind, lineno: int) -> bool:
             if style.overline is not None:
                 memo.section_bubble_up_kludge = True
 
+            if level <= self.memo.minimum_level_stack[-1]:
+                self.parent.append(
+                    self.reporter.error(
+                        "Block content cannot contain sections that are sibling to the parent element's section level",
```
Comment (schmalliso) [inline snooty/tinydocutils/states.py:1115]: So I understand what this message means, but I'm not sure our users will. Can we provide an example of when this happens and how to remediate?

File: snooty/test_parser.py [hunk 2]
```diff
@@ -3730,3 +3730,133 @@ def test_duplicate_directive_option() -> None:
...
+    """DOP-3969"""
+    path = FileId("test.rst")
+    project_config = ProjectConfig(ROOT_PATH, "")
+    parser = rstparser.Parser(project_config, JSONVisitor)
+
+    page, diagnostics = parse_rst(
+        parser,
+        path,
+        """
+Query Result Format
```
Comment (schmalliso) [inline snooty/test_parser.py:3745]: Can you add some comments to these describing why they're wonky or not wonky so it's clearer to scan?

**Review by schmalliso — CHANGES_REQUESTED** [review]
Looks good to me structurally, just a few tiny suggestions to enhance clarity.

File: snooty/tinydocutils/states.py (hunk 1 above)
Comment (i80and) [reply snooty/tinydocutils/states.py:1115]: It's tricky -- I would like to have a more detailed error message with an example, but historically we have short single-line error messages, with an exception for the rare suggested correction, and any explanation will be vertically very long.

What do you think about a link to a page to explain the issue in more detail, perhaps?

Comment (schmalliso) [reply snooty/tinydocutils/states.py:1115]: That sounds *great*!

Comment (i80and) [reply snooty/tinydocutils/states.py:1115]: https://github.com/mongodb/snooty-parser/wiki/Section-Hierarchy-Must-Be-Linear

How does this look?

File: snooty/test_parser.py (hunk 2 above)
Comment (i80and) [reply snooty/test_parser.py:3746]: Done! ✍🏻

**Review by schmalliso — APPROVED** [review]

**Pull request merged**